js/core/telemetrySyncManager.js

```diff
@@ -13,56 +13,83 @@ var TelemetrySyncManager = {
      * @memberof TelemetryPlugin
      */
     _teleData: [],
+    _failedBatch: [],
+    _failedBatchSize: 20,
+    _syncRetryInterval: 2000,
     init: function() {
         var instance = this;
+        var Telemetry = EkTelemetry || Telemetry;
+        Telemetry.config.syncRetryInterval && (instance._syncRetryInterval = Telemetry.config.syncRetryInterval);
+        Telemetry.config.failedBatchSize && (instance._failedBatchSize = Telemetry.config.failedBatchSize);
         document.addEventListener('TelemetryEvent', this.sendTelemetry);
     },
     sendTelemetry: function(event) {
         var telemetryEvent = event.detail;
-        console.log("Telemetry Events ", JSON.stringify(telemetryEvent));
+        Telemetry.config.telemetryDebugEnabled && console.log("Telemetry Events ", JSON.stringify(telemetryEvent));
         var instance = TelemetrySyncManager;
         instance._teleData.push(Object.assign({}, telemetryEvent));
         if ((telemetryEvent.eid.toUpperCase() === "END") || (instance._teleData.length >= Telemetry.config.batchsize)) {
             TelemetrySyncManager.syncEvents();
         }
     },
-    updateEventStack: function(events) {
-        TelemetrySyncManager._teleData = TelemetrySyncManager._teleData.concat(events);
-    },
-    syncEvents: function() {
+    syncEvents: function(telemetryObj) {
         var Telemetry = EkTelemetry || Telemetry;
         var instance = TelemetrySyncManager;
-        var telemetryData = instance._teleData.splice(0, Telemetry.config.batchsize);
-        var telemetryObj = {
-            "id": "ekstep.telemetry",
-            "ver": Telemetry._version,
-            "ets": (new Date()).getTime(),
-            "events": telemetryData
-        };
+        if(!telemetryObj){
+            var telemetryEvents = instance._teleData.splice(0, Telemetry.config.batchsize);
+            if(!telemetryEvents.length){
+                return;
+            }
+            telemetryObj = {
+                "id": "ekstep.telemetry",
+                "ver": Telemetry._version,
+                "params": {
+                    "msgid": CryptoJS.MD5(JSON.stringify(telemetryEvents)).toString(),
+                },
+                "ets": (new Date()).getTime(),
```

Comment:
The "ets" logic should be similar to telemetry event ets logic where we use the server timestamp to set the right date on the client system.